File: 动态规划/python/isMatch.py
"""
正则匹配，给定字符串s和p。实现支持" . "和" * "的正则匹配
" . "：匹配单个字符
" * "：匹配零个或多个前面的字符

------------------------------------------------

例1：输入 s = "aa", p = "a*", 返回True
例2：输入 s = "aab", p = "c*a*b", 返回True
例3：输入 s = "ab", p = ".*", 返回True
"""


def isMatch(text, pattern):
    if len(pattern) == 0:
        return len(text) == 0

    first_match = len(text) > 0 and pattern[0] in [".", text[0]]
    if len(pattern) >= 2 and pattern[1] == "*":
        # 匹配0次，即pattern右移2位
        # 匹配1次，即text右移1位，666
        return isMatch(text, pattern[2:]) or (first_match and isMatch(text[1:], pattern))
    else:
        return first_match and isMatch(text[1:], pattern[1:])


def isMatchMemo(text, pattern):
    d = {}
    def dp(i, j):
        if (i, j) in d:
            return d[(i, j)]
        if j == len(pattern):
            return len(text) == i

        first_match = len(text[i:]) > 0 and pattern[j] in [".", text[i]]
        if len(pattern[j:]) >= 2 and pattern[j+1] == "*":
            ans = dp(i, j+2) or (first_match and dp(i+1, j))

        else:
            ans = first_match and dp(i+1, j+1)
        d[(i, j)] = ans
        return ans
    return dp(0, 0)


if __name__ == "__main__":
    s = "aaabc"
    p = "a.*bcd"
    print(isMatchMemo(s, p))

File: 动态规划/cpp/isMatch.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>

using namespace std;

struct double_string
{
    string text;
    string pat;
    double_string(string _text="", string _pat="") : text(_text), pat(_pat) {}
    bool operator==(const double_string& ds) const {return (text==ds.text && pat==ds.pat);}
    friend ostream& operator<<(ostream& os, const double_string& ds)
    {
        os << "text: " << ds.text << ", pat: " << ds.pat << endl;
        return os;
    }
};

namespace std
{
    template<>
    struct hash<double_string>
    {
        size_t operator() (const double_string& s) const noexcept
        {
            return  hash<decltype(s.text)>()(s.text) +
                    hash<decltype(s.pat)>()(s.pat);
        }
    }; // 间接调用原生Hash.
}



bool isMatch(const string& text, const string& pat)
{   
    int M = text.size();
    int N = pat.size();
    if(N==0)
        return M==0;
    bool first_match = (text[0] == pat[0] || pat[0] == '.');

    // *匹配0次或匹配1次
    if(N >= 2 && pat[1]== '*')
        return isMatch(text, pat.substr(2, N-2)) || (first_match && isMatch(text.substr(1, M-1), pat.substr(0, N)));
    else
        return first_match && isMatch(text.substr(1, M-1), pat.substr(1, N-1));
}

bool isMatchMemo(const string& text, const string& pat, unordered_map<double_string, bool>& memo)
{   
    int M = text.size();
    int N = pat.size();

    if(N==0)
        return M==0;
    
    double_string tmp(text, pat);
    if(memo.count(tmp))
        return memo[tmp];

    bool first_match = (text[0] == pat[0] || pat[0] == '.');
    bool res;
    // *匹配0次或匹配1次
    if(N >= 2 && pat[1]== '*')
        res = isMatch(text, pat.substr(2, N-2)) || (first_match && isMatch(text.substr(1, M-1), pat.substr(0, N)));
    else
        res = first_match && isMatch(text.substr(1, M-1), pat.substr(1, N-1));
    memo[tmp] = res;
    return res;
}

int main()
{
    string text = "aaa";
    string pat = "a*";
    unordered_map<double_string, bool> memo;
    cout << isMatchMemo(text, pat, memo) << endl;
    return 0;
}
